component/file_cache/file_cache.go

```diff
@@ -466,19 +490,70 @@ func (fc *FileCache) StreamDir(options internal.StreamDirOptions) ([]*internal.O
 	// 3. Path in storage and in local cache (this could result in dirty properties on the service if we recently wrote to the file)
 
 	// To cover case 1, grab all entries from storage
+
+	// If the cloud is down, we get stale data back from the attribute cache
+	log.Debug("FileCache::StreamDir : the currFile name is %s", options.Name)
 	attrs, token, err := fc.NextComponent().StreamDir(options)
 	if err != nil {
-		return attrs, token, err
+		token = ""
+		var maxAttempts *retry.MaxAttemptsError
+		cloudIsDown := errors.As(err, &maxAttempts)
+		if cloudIsDown {
+			err = nil
+		}
+	}
+
+	//after getting the stale cache listing, compare it with async map so that it's correct
+	//this is at a high performance cost, but can be optimized later
+	i := 0
+	for i < len(attrs) {
+
+		fc.fileOps.Range(func(key, value interface{}) bool {
+			val := value.(FileAttributes)
+			keyString := key.(string)
+			attr := attrs[i]
+
+			if val.operation == "RenameFile" || val.operation == "RenameDir" { //for rename operations, only grab the src file/dir to compare
+
+				keyString = keyString[:strings.IndexByte(keyString, ',')]
+
+			}
+
+			if attr.Name == keyString {
+				//remove file listing if there is a deleteFile operation
+				if val.operation == "DeleteFile" {
+					attrs = append(attrs[:i], attrs[i+1:]...)
+				}
+				//Change source file name to dst file name
+				if val.operation == "RenameFile" {
+					renameOptions := val.options.(internal.RenameFileOptions)
+					attrs[i].Name = renameOptions.Dst //this just does a rename but we need to make sure file contents are also transferred
+				}
```

Comment:
Since rename can also move a file (or folder), you might need to check if the file still belongs in the this directory listing after its rename.